### ISSUE TITLE
Endless "datum already been processed" looping

### ISSUE THREAD

**Issue by dwhitena** [pachyderm/pachyderm#1674]
# Behavior

1. A pipeline is create, re-created, or updated (this has been seen for the first run of a pipeline, for runs after a failure, or when a pipeline is updated).
2. The corresponding job starts running.
3. The job starts to restart as indicated by the restart ticker from `list-job`, and it never ends.
4. The non-raw logs from the job indicate that the code ran like normal through the datums.
5. The raw logs from the job indicate that the job is constantly being restarted and looping as follows endlessly:

```
[{"path":"/013_CRa_H9M5_M470_Blank_04_alternate_pos_low_mr.mzML","hash":"5Hk/YSV799tZMfJKlB8mv7hiEUH4ttm/CPSx42HUkLI="}],"ts":"2017-04-12T13:32:31.458187621Z","message":"Received request"}
{"pipelineName":"newpipeline","pipelineId":"d2514a890d3a4a84b82a6a4eb015c92a","jobId":"ee02de42-b24f-4aa4-892a-720de35f0d45","workerId":"pipeline-newpipeline-v1-x12t7","data":[{"path":"/013_CRa_H9M5_M470_Blank_04_alternate_pos_low_mr.mzML","hash":"5Hk/YSV799tZMfJKlB8mv7hiEUH4ttm/CPSx42HUkLI="}],"ts":"2017-04-12T13:32:31.461723545Z","message":"skipping input, as it's already been processed"}
{"pipelineName":"newpipeline","pipelineId":"d2514a890d3a4a84b82a6a4eb015c92a","jobId":"ee02de42-b24f-4aa4-892a-720de35f0d45","workerId":"pipeline-newpipeline-v1-x12t7","data":[{"path":"/019_CRa_H9M5_M470_C01_K3_alternate_pos_low_mr.mzML","hash":"tSxgLrLiJFVwTrs+POCEJBXnYmH65O2P+/sHLEFEJdo="}],"ts":"2017-04-12T13:32:31.466704655Z","message":"Received request"}
{"pipelineName":"newpipeline","pipelineId":"d2514a890d3a4a84b82a6a4eb015c92a","jobId":"ee02de42-b24f-4aa4-892a-720de35f0d45","workerId":"pipeline-newpipeline-v1-x12t7","data":[{"path":"/019_CRa_H9M5_M470_C01_K3_alternate_pos_low_mr.mzML","hash":"tSxgLrLiJFVwTrs+POCEJBXnYmH65O2P+/sHLEFEJdo="}],"ts":"2017-04-12T13:32:31.468799930Z","message":"skipping input, as it's already been processed"}
etc...
```

**Note** - This seems to happen most in cases where we are processing larger files (100Mb or more), and it seems to only happen when processing multiple files (i.e., multiple files in one commit).

**Note** - This also seems to happen more often with higher numbers of nodes in the cluster (e.g., 6 instead of 3).

**Note** - This seems to happen particularly often with Minio clusters.  However, I'm pretty sure @sjezewski  has seen it for normal clusters.  Thus, I don't think it is a Minio specific issue.  However, I do think Minio related bugs, upload/download behavior somehow exacerbates the issue.  Possibly because we are working with large files.

# Steps to reproduce

I will illustrate here with a GKE cluster, as I have most experience here.  But, afaik, this is restricted to GKE because the users reporting it had other deployments.

Also, I will demonstrate this with Pachyderm built from master (at least built from master yesterday).  It's probably good to use the master images of whatever the next release is after 1.4.3, because the zombie jobs bugs might not let you see the below output.

## 1. Deploy a Minio + Pachyderm cluster

You can either run Minio in k8s or just create a Minio standalone instance.  Assuming you want to use a standalone instance, create an instance in Google cloud (with docker) and run the following to start Minio:

```
docker run -d -p 9000:9000 minio/minio server /export
```

Also, make sure that port 9000 is exposed.  The Minio creds (ID, secret, etc.) that you will need when deploying pachyderm can be found by:

```
docker logs <container id>
```

And you can create a bucket for pachyderm easily via the Minio web browser (accessible on port 9000), or with the `mc` Minio client.

## 2. Deploy Pachyderm + Minio

```
➔ CLUSTER_NAME=pach-cluster
➔ MACHINE_TYPE=n1-standard-2
➔ gcloud container clusters create ${CLUSTER_NAME} --scopes storage-rw --machine-type ${MACHINE_TYPE} --num-nodes=6
Creating cluster pach-cluster...done.                                                                                                                                                   
Created [https://container.googleapis.com/v1/projects/dan-dev-155821/zones/us-west1-a/clusters/pach-cluster].
kubeconfig entry generated for pach-cluster.
NAME          ZONE        MASTER_VERSION  MASTER_IP        MACHINE_TYPE   NODE_VERSION  NUM_NODES  STATUS
pach-cluster  us-west1-a  1.5.6           104.199.126.191  n1-standard-2  1.5.6         6          RUNNING
➔ gcloud container clusters get-credentials ${CLUSTER_NAME}
Fetching cluster endpoint and auth data.
kubeconfig entry generated for pach-cluster.
➔ STORAGE_SIZE=10
➔ STORAGE_NAME=pach-minio-disk
➔ gcloud compute disks create --size=${STORAGE_SIZE}GB ${STORAGE_NAME}
WARNING: You have selected a disk size of under [200GB]. This may result in poor I/O performance. For more information, see: https://developers.google.com/compute/docs/disks#pdperformance.
Created [https://www.googleapis.com/compute/v1/projects/dan-dev-155821/zones/us-west1-a/disks/pach-minio-disk].
NAME             ZONE        SIZE_GB  TYPE         STATUS
pach-minio-disk  us-west1-a  10       pd-standard  READY

New disks are unformatted. You must format and mount a disk before it
can be used. You can find instructions on how to do this at:

https://cloud.google.com/compute/docs/disks/add-persistent-disk#formatting

➔ pachctl deploy custom --persistent-disk google --object-store s3 ${STORAGE_NAME} ${STORAGE_SIZE} <minio bucket> <minio ID> <minio secret> <minio IP>:9000 --static-etcd-volume=${STORAGE_NAME}
serviceaccount "pachyderm" created
persistentvolume "etcd-volume" created
persistentvolumeclaim "etcd-storage" created
replicationcontroller "etcd" created
service "etcd" created
service "pachd" created
replicationcontroller "pachd" created
secret "minio-secret" created
➔ kubectl get all
NAME             READY     STATUS    RESTARTS   AGE
po/etcd-ps0jg    1/1       Running   0          3m
po/pachd-9zbl0   1/1       Running   5          3m

NAME       DESIRED   CURRENT   READY     AGE
rc/etcd    1         1         1         3m
rc/pachd   1         1         1         3m

NAME             CLUSTER-IP     EXTERNAL-IP   PORT(S)                       AGE
svc/etcd         10.3.250.175   <nodes>       2379:31104/TCP                3m
svc/kubernetes   10.3.240.1     <none>        443/TCP                       16m
svc/pachd        10.3.249.84    <nodes>       650:30650/TCP,651:30651/TCP   3m
➔ 
```

## 3. Commit multiple large files

All of the data referenced below is publicly available:

```
wget ftp://ftp.ebi.ac.uk/pub/databases/metabolights/studies/public/MTBLS233/*alternate_pos_low_mr.mzML
```

In my case, I only had to use 4 of these files to reproduce the issue, but again it seems to be particular tied to multiple files. 

```
$ ls
001_CRa_H9M5_M470_Blank_01_alternate_pos_low_mr.mzML  003_CRa_H9M5_M470_Pool_02_alternate_pos_low_mr.mzML
002_CRa_H9M5_M470_Pool_01_alternate_pos_low_mr.mzML   005_CRa_H9M5_M470_Blank_02_alternate_pos_low_mr.mzML
$ pachctl create-repo metabrepotest
$ pachctl put-file metabrepotest master -c -r -f .
$ pachctl list-repo
NAME                CREATED              SIZE                
metabrepotest       About a minute ago   420.4 MiB           
$ pachctl list-file metabrepotest master
NAME                                                   TYPE                SIZE                
001_CRa_H9M5_M470_Blank_01_alternate_pos_low_mr.mzML   file                107.7 MiB           
002_CRa_H9M5_M470_Pool_01_alternate_pos_low_mr.mzML    file                104.1 MiB           
003_CRa_H9M5_M470_Pool_02_alternate_pos_low_mr.mzML    file                104.8 MiB           
005_CRa_H9M5_M470_Blank_02_alternate_pos_low_mr.mzML   file                103.7 MiB           
$
```

## 4. Create the pipeline

The pipeline spec can be found here: https://www.dropbox.com/s/ldeb3gn6f49rjju/PeakPickerHiRes.json

```
$ pachctl create-pipeline -f PeakPickerHiRes.json 
$ pachctl list-pipeline
NAME                INPUT               OUTPUT                   STATE               
PeakPickerHiRes     metabrepotest       PeakPickerHiRes/master   running    
$ pachctl list-job
ID                                   OUTPUT COMMIT     STARTED       DURATION RESTART PROGRESS STATE            
f6df06a0-cf08-4969-918a-9479abe71610 PeakPickerHiRes/- 7 seconds ago -        0       0 / 4    running 
$
```

## Check out what happens

As you will notice, the job starts to restart endlessly:

```
$ pachctl create-pipeline -f PeakPickerHiRes.json 
$ pachctl list-job
ID                                   OUTPUT COMMIT     STARTED       DURATION RESTART PROGRESS STATE            
66dc98ed-f03d-4ea6-af39-6537e24cf8c4 PeakPickerHiRes/- 4 seconds ago -        0       0 / 4    running 
$ pachctl list-job
ID                                   OUTPUT COMMIT     STARTED       DURATION RESTART PROGRESS STATE            
66dc98ed-f03d-4ea6-af39-6537e24cf8c4 PeakPickerHiRes/- 6 seconds ago -        0       0 / 4    running 
$ pachctl list-job
ID                                   OUTPUT COMMIT     STARTED       DURATION RESTART PROGRESS STATE            
66dc98ed-f03d-4ea6-af39-6537e24cf8c4 PeakPickerHiRes/- 8 seconds ago -        0       0 / 4    running 
$ pachctl list-job
ID                                   OUTPUT COMMIT     STARTED        DURATION RESTART PROGRESS STATE            
66dc98ed-f03d-4ea6-af39-6537e24cf8c4 PeakPickerHiRes/- 10 seconds ago -        1       4 / 4    running 
$ pachctl list-job
ID                                   OUTPUT COMMIT     STARTED        DURATION RESTART PROGRESS STATE            
66dc98ed-f03d-4ea6-af39-6537e24cf8c4 PeakPickerHiRes/- 12 seconds ago -        3       4 / 4    running 
$ pachctl list-job
ID                                   OUTPUT COMMIT     STARTED        DURATION RESTART PROGRESS STATE            
66dc98ed-f03d-4ea6-af39-6537e24cf8c4 PeakPickerHiRes/- 13 seconds ago -        4       4 / 4    running 
$ pachctl list-job
ID                                   OUTPUT COMMIT     STARTED        DURATION RESTART PROGRESS STATE            
66dc98ed-f03d-4ea6-af39-6537e24cf8c4 PeakPickerHiRes/- 14 seconds ago -        4       4 / 4    running 
$ pachctl list-job
ID                                   OUTPUT COMMIT     STARTED        DURATION RESTART PROGRESS STATE            
66dc98ed-f03d-4ea6-af39-6537e24cf8c4 PeakPickerHiRes/- 16 seconds ago -        5       4 / 4    running 
$
```

If we look at the non-raw logs we see:

```
$ pachctl get-logs --job 66dc98ed-f03d-4ea6-af39-6537e24cf8c4
$
```

Which tells us our code didn't run or process anything.  But then if we look at the raw logs we can see:

```
$ pachctl get-logs --job 66dc98ed-f03d-4ea6-af39-6537e24cf8c4 --raw
{"pipelineName":"PeakPickerHiRes","pipelineId":"2f42ca6593a34497ade0ee4db0fe8262","jobId":"66dc98ed-f03d-4ea6-af39-6537e24cf8c4","workerId":"pipeline-peakpickerhires-v1-370fn","data":[{"path":"/001_CRa_H9M5_M470_Blank_01_alternate_pos_low_mr.mzML","hash":"pS3X+QOC554kGT1wjPRH8iUu7G2QI+67199mmCWt+aI="}],"ts":"2017-04-14T21:33:05.184130161Z","message":"Received request"}
{"pipelineName":"PeakPickerHiRes","pipelineId":"2f42ca6593a34497ade0ee4db0fe8262","jobId":"66dc98ed-f03d-4ea6-af39-6537e24cf8c4","workerId":"pipeline-peakpickerhires-v1-370fn","data":[{"path":"/001_CRa_H9M5_M470_Blank_01_alternate_pos_low_mr.mzML","hash":"pS3X+QOC554kGT1wjPRH8iUu7G2QI+67199mmCWt+aI="}],"ts":"2017-04-14T21:33:05.191463108Z","message":"skipping input, as it's already been processed"}
{"pipelineName":"PeakPickerHiRes","pipelineId":"2f42ca6593a34497ade0ee4db0fe8262","jobId":"66dc98ed-f03d-4ea6-af39-6537e24cf8c4","workerId":"pipeline-peakpickerhires-v1-370fn","data":[{"path":"/002_CRa_H9M5_M470_Pool_01_alternate_pos_low_mr.mzML","hash":"tvOz8tDu8gRNJN5R8yCL2Q4Bxzp3W3Xi4dVXXBZiMas="}],"ts":"2017-04-14T21:33:06.235654927Z","message":"Received request"}
{"pipelineName":"PeakPickerHiRes","pipelineId":"2f42ca6593a34497ade0ee4db0fe8262","jobId":"66dc98ed-f03d-4ea6-af39-6537e24cf8c4","workerId":"pipeline-peakpickerhires-v1-370fn","data":[{"path":"/002_CRa_H9M5_M470_Pool_01_alternate_pos_low_mr.mzML","hash":"tvOz8tDu8gRNJN5R8yCL2Q4Bxzp3W3Xi4dVXXBZiMas="}],"ts":"2017-04-14T21:33:06.238075317Z","message":"skipping input, as it's already been processed"}
{"pipelineName":"PeakPickerHiRes","pipelineId":"2f42ca6593a34497ade0ee4db0fe8262","jobId":"66dc98ed-f03d-4ea6-af39-6537e24cf8c4","workerId":"pipeline-peakpickerhires-v1-370fn","data":[{"path":"/003_CRa_H9M5_M470_Pool_02_alternate_pos_low_mr.mzML","hash":"CMykQv2SCTniKAJvLsM3H3SnFuk3Xlx4e1/1SaGqOXM="}],"ts":"2017-04-14T21:33:10.949767432Z","message":"Received request"}
{"pipelineName":"PeakPickerHiRes","pipelineId":"2f42ca6593a34497ade0ee4db0fe8262","jobId":"66dc98ed-f03d-4ea6-af39-6537e24cf8c4","workerId":"pipeline-peakpickerhires-v1-370fn","data":[{"path":"/003_CRa_H9M5_M470_Pool_02_alternate_pos_low_mr.mzML","hash":"CMykQv2SCTniKAJvLsM3H3SnFuk3Xlx4e1/1SaGqOXM="}],"ts":"2017-04-14T21:33:10.951892334Z","message":"skipping input, as it's already been processed"}
etc...
```

Which in some cases seems to continue forever. Actually, in this particular case (the one I'm reproducing here), the job does eventually end.  However, it takes a lot longer than it should and includes 18 restarts:

```
dwhitena@instance-1:~$ pachctl list-job
ID                                   OUTPUT COMMIT                                    STARTED       DURATION  RESTART PROGRESS STATE            
66dc98ed-f03d-4ea6-af39-6537e24cf8c4 PeakPickerHiRes/872130f9611740d58f7b30c2f4210413 4 minutes ago 3 minutes 18      4 / 4    success
```

In the case of many more files, it seems like this looping can really block a job and cause it to stall for very long periods of time.  As is illustrated above, maybe if you wait for those long periods of time, the job may complete, but this isn't really practical for many use cases.

For this particular case.  Here are the pachd logs:

[pachd.txt](https://github.com/pachyderm/pachyderm/files/923003/pachd.txt)

# Notes

Again, this seems to not also result in true "endless" looping (i.e., that goes on for infinity), because the above example job did end up finishing.  However, there are a ton of restarts which are troubling, and, for larger numbers of files, this seems to be a total blocker.

@sjezewski feel free to add any details about the scenarios in which you have seen users getting into this situation.  

Thanks!

**Comment by msteffen** [comment]
Thanks for the incredibly detailed bug! Right now, I think pachyderm will enter this loop any time a job fails for any reason. We restart jobs indefinitely, and each time we do, the job will try to reprocess all data in the input commit (including things that have already been processed successfully). Thus if `CMykQv2SCTniKAJvLsM3H3SnFuk3Xlx4e1/1SaGqOXM=` is processed successfully in the first run, you'll see the message above printed in a loop, until the job completes successfully.

**Comment by derekchiang** [comment]
To clarify, if the user code fails more than 3 times, the job is marked as `failure` and thus not retired.  From the logs that @dwhitena posted, it looks like the job is failing due to non-user-related reasons.  Specifically, the job seems to have trouble retrieving certain objects that it thinks exist.

**Comment by derekchiang** [comment]
The minio guys were kind enough to provide the following info:

```
...looks like an object is being read before the multi-part upload is completed. 

Basically, in a multi-part upload, there are three steps involved - initMultipart upload, actual upload of parts, and finalisation of the uploaded object. We observed that initMultipart is done, but before the finalization, there are multiple attempts to read the object. In the end upload is being finalized, but there are no read attempts after that.
```

**Comment by derekchiang** [comment]
Fixed by #1723
